### PR TITLE
Visitor can sign-up, login and logout

### PR DESCRIPTION
**Pull request by Carrosen**
# [PT Story](https://www.pivotaltracker.com/story/show/166572555)

## Changes we propose in this PR
* Installs and configures Redux-token-auth
* Implementing axios interceptor
* Adds restrictions to make sure only registered users can create an entry
* Basic styling to log-in and sign-up forms

## What we have learned
* More about redux-token-auth
* How to restrict and redirect depending on if user is logged in

## Screenshots
Screenshot from 2019-06-21 12-54-12

Screenshot from 2019-06-21 12-53-55

### CODE REVIEW

File: cypress/integration/VisitorCanCreateAPostUnsuccessfully.spec.js [hunk 1]
```diff
@@ -8,8 +8,31 @@ describe('Visitor can', () => {
...
     cy.visit('http://localhost:3000')
     cy.get('#map-icon-plus').click()
+    cy.get('#login-form').within(() => {
+      cy.get('#email').type('carla@mail.com')
+      cy.get('#password').type('password')
+    })
+    cy.get('#login_form_button').click()
+    cy.contains('You have succesfully logged in')
+    cy.wait(3000)
+    cy.get('#map-icon-plus').click()
```
Comment (raptorf1) [inline cypress/integration/VisitorCanCreateAPostUnsuccessfully.spec.js:35]: Comment is for lines 42-44. But I cannot write it further down.
Maybe make a loop for those `cy.contains`?

File: cypress/integration/visitorCanSignUp.spec.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+      response: 'fixture:successful_signup.json'
+    })
+    cy.visit('http://localhost:3000')
+    cy.get('#profile-icon').click()
+    cy.get('#sign_up_link').click()
+    cy.get('#signup-form').within(() => {
+      cy.get('#email').type('zane@mail.com')
+      cy.get('#password').type('password')
+      cy.get('#password_confirmation').type('password')
+    })
```
Comment (raptorf1) [inline cypress/integration/visitorCanSignUp.spec.js]: a loop for those `cy.get` and `type`?

File: cypress/integration/visitorCanSignUp.spec.js [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+      response: 'fixture:unsuccessful_signup.json',
+    })
+    cy.visit('http://localhost:3000')
+    cy.get('#profile-icon').click()
+    cy.get('#sign_up_link').click()
+    cy.get('#signup-form').within(() => {
+      cy.get('#email').type('boa@')
+      cy.get('#password').type('pass')
+      cy.get('#password_confirmation').type('password')
+    })
```
Comment (raptorf1) [inline cypress/integration/visitorCanSignUp.spec.js:61]: loop?

File: cypress/integration/visitorCanSignUp.spec.js [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    cy.get('#sign_up_button').click()
+    cy.contains("Password confirmation doesn't match Password")
+    cy.contains("Password is too short (minimum is 6 characters")
+    cy.contains("Email is not an email")
+    cy.contains("Level can't be blank")
+  })
```
Comment (raptorf1) [inline cypress/integration/visitorCanSignUp.spec.js]: and again loop :)

File: src/Components/Login.jsx [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+    e.preventDefault();
+    const { history, signInUser } = this.props
+    const {
+      email,
+      password
+    } = this.state
+    signInUser({ email, password })
+      .then(response => {
+        this.setState({ message: true })
+        setTimeout(function () { history.push('/') }, 3000)
```
Comment (raptorf1) [inline src/Components/Login.jsx]: maybe do that 1000 like in Glocal? or 1500

File: src/Components/Map.jsx [hunk 6]
```diff
@@ -1,11 +1,12 @@
...
 import GoogleMapReact from 'google-map-react';
 import MapStyle from '../Modules/MapStyle'
 import { Icon } from 'semantic-ui-react';
 import Popup from 'reactjs-popup';
-import CreateImageEntry from './CreateImageEntry'
-import axios from 'axios'
-import { connect } from 'react-redux'
+import CreateImageEntry from './CreateImageEntry';
+import axios from 'axios';
+import { connect } from 'react-redux';
```
Comment (raptorf1) [inline src/Components/Map.jsx]: Let's be consistent with these `;`. We either keep them or we dump them. In other Components we don't use them. Personally I vote for dumping them. But whatever we decide, we stick with it all over the app.

File: src/Components/SignUp.jsx [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+    const {
+      email,
+      password,
+      password_confirmation,
+      level,
+    } = this.state
+    registerUser({ email, password, password_confirmation, level })
+      .then(response => {
+        this.setState({ message: true })
+        setTimeout(function () { history.push('/') }, 3000)
```
Comment (raptorf1) [inline src/Components/SignUp.jsx]: same comment as in Login component. maybe do this 1000 or 1500?

File: src/index.js [hunk 8]
```diff
@@ -7,15 +7,21 @@ import { BrowserRouter } from 'react-router-dom'
...
+import { setupInterceptors } from './Modules/axiosInterceptor';
+import { verifyCredentials } from './reduxTokenAuthConfig'
+
+setupInterceptors()
 
 const loc = 'http://localhost:3002'
 const her = 'https://being-stockholm.herokuapp.com'
 
-axios.defaults.baseURL = her
+axios.defaults.baseURL = loc
```
Comment (raptorf1) [inline src/index.js]: change back to Heroku before we merge

File: src/reduxTokenAuthConfig.js [hunk 9]
```diff
@@ -0,0 +1,26 @@
+import { generateAuthActions } from 'redux-token-auth'
+
+const config = {
+  authUrl: "http://localhost:3002/api/v1/auth",
```
Comment (raptorf1) [inline src/reduxTokenAuthConfig.js]: do here the same we do in the App.js with the `loc` and `her` variables. and use single quotes.

**Review by raptorf1 — CHANGES_REQUESTED** [review]

File: cypress/integration/VisitorCanCreateAPostUnsuccessfully.spec.js (hunk 1 above)
Comment (stefankarlberg) [reply cypress/integration/VisitorCanCreateAPostUnsuccessfully.spec.js:35]: Fixed!

File: cypress/integration/visitorCanSignUp.spec.js (hunk 2 above)
Comment (stefankarlberg) [reply cypress/integration/visitorCanSignUp.spec.js]: Fixed!

File: cypress/integration/visitorCanSignUp.spec.js (hunk 3 above)
Comment (stefankarlberg) [reply cypress/integration/visitorCanSignUp.spec.js:61]: Fixed!

File: cypress/integration/visitorCanSignUp.spec.js (hunk 4 above)
Comment (stefankarlberg) [reply cypress/integration/visitorCanSignUp.spec.js]: Fixed!

File: src/Components/Login.jsx (hunk 5 above)
Comment (stefankarlberg) [reply src/Components/Login.jsx]: Fixed!

File: src/Components/Map.jsx (hunk 6 above)
Comment (stefankarlberg) [reply src/Components/Map.jsx]: Fixed!

File: src/Components/SignUp.jsx (hunk 7 above)
Comment (stefankarlberg) [reply src/Components/SignUp.jsx]: Fixed!

File: src/reduxTokenAuthConfig.js (hunk 9 above)
Comment (stefankarlberg) [reply src/reduxTokenAuthConfig.js]: Fixed!

**Pull request merged**